File: packages/jelly_tabs/lib/src/math/pill_jelly_animation.dart
import 'dart:math' as math;

import 'package:jelly_tabs/src/config/config.dart';
import 'package:jelly_tabs/src/math/animation_math.dart';

/// Mutable state for the pill jelly frame loop, mirroring
/// Reanimated shared values in `pill-jelly-animation.ts`.
class PillJellyFrameState {
  /// Horizontal pill scale, spring-driven toward [shapeTarget].
  double baseScaleX = 1;

  /// Velocity of [baseScaleX].
  double baseScaleXRate = 0;

  /// Vertical pill scale, spring-driven toward [shapeTarget].
  double baseScaleY = 1;

  /// Velocity of [baseScaleY].
  double baseScaleYRate = 0;

  /// Filtered pill velocity used for velocity-shear correction.
  double filteredVelocity = 0;

  /// Velocity of [filteredVelocity].
  double filteredVelocityRate = 0;

  /// 1 while a gesture is active, 0 otherwise.
  double isDragging = 0;

  /// Press inflation progress (0 → 1).
  double pressProgress = 0;

  /// Velocity of [pressProgress].
  double pressProgressRate = 0;

  /// The press-inflation target (1 while pressed, 0 on release).
  double pressTarget = 0;

  /// The raw, unclamped panel horizontal offset.
  double rawPanelOffset = 0;

  /// Velocity of [rawPanelOffset].
  double rawPanelOffsetVelocity = 0;

  /// 1 while a settle is pending after release, 0 otherwise.
  double releasePending = 0;

  /// The pill's scale target (1 at rest, `pressedScale` while pressed).
  double shapeTarget = 1;

  /// The target tab position (0..maxTabIndex).
  double targetValue = 0;

  /// The pill's current tab position.
  double value = 0;

  /// Velocity of [value].
  double valueVelocity = 0;
}

/// Advances the pill jelly frame state by one step, mirroring
/// `advancePillJellyFrame` in `pill-jelly-animation.ts`.
void advancePillJellyFrame(
  PillJellyFrameState state,
  PillJellyFrameConfig config, {
  required int tabCount,
  required int dtMs,
}) {
  final dt = getFrameDeltaSeconds(dtMs);
  if (dt == null) return;

  final maxIndex = getMaxTabIndex(tabCount);
  state.targetValue = state.targetValue.clamp(0.0, maxIndex.toDouble());

  final valueStep = advanceSpring(
    state.value,
    state.valueVelocity,
    state.targetValue,
    spring: config.springs.value,
    dt: dt,
  );
  state
    ..value = valueStep.value
    ..valueVelocity = valueStep.velocity;

  final velocityTarget = (state.isDragging > 0 && maxIndex > 0)
      ? state.valueVelocity / maxIndex
      : 0.0;
  final velocityStep = advanceSpring(
    state.filteredVelocity,
    state.filteredVelocityRate,
    velocityTarget,
    spring: config.springs.velocity,
    dt: dt,
  );
  state
    ..filteredVelocity = velocityStep.value
    ..filteredVelocityRate = velocityStep.velocity;

  // rawPanelOffset only springs back when not dragging
  if (state.isDragging == 0) {
    final panelStep = advanceSpring(
      state.rawPanelOffset,
      state.rawPanelOffsetVelocity,
      0,
      spring: config.springs.panel,
      dt: dt,
    );
    state
      ..rawPanelOffset = panelStep.value
      ..rawPanelOffsetVelocity = panelStep.velocity;
  }

  // settle released indicator
  if (state.releasePending > 0) {
    final releaseThreshold =
        config.releaseDistanceFraction * math.max(1, maxIndex);
    if ((state.value - state.targetValue).abs() < releaseThreshold) {
      state
        ..releasePending = 0
        ..pressTarget = 0
        ..shapeTarget = 1;
    }
  }

  final pressStep = advanceSpring(
    state.pressProgress,
    state.pressProgressRate,
    state.pressTarget,
    spring: config.springs.press,
    dt: dt,
  );
  state
    ..pressProgress = pressStep.value
    ..pressProgressRate = pressStep.velocity;

  final scaleXStep = advanceSpring(
    state.baseScaleX,
    state.baseScaleXRate,
    state.shapeTarget,
    spring: config.springs.scaleX,
    dt: dt,
  );
  state
    ..baseScaleX = scaleXStep.value
    ..baseScaleXRate = scaleXStep.velocity;

  final scaleYStep = advanceSpring(
    state.baseScaleY,
    state.baseScaleYRate,
    state.shapeTarget,
    spring: config.springs.scaleY,
    dt: dt,
  );
  state
    ..baseScaleY = scaleYStep.value
    ..baseScaleYRate = scaleYStep.velocity;
}

File: reference/react-native-jelly-tabs/src/utils/pill-jelly-animation.ts
import {
    advanceSharedSpring,
    getFrameDeltaSeconds,
    getMaxTabIndex,
    type SpringConfig,
} from "./animation";
import type { SharedValue } from "react-native-reanimated";

type SharedNumber = SharedValue<number>;

export interface PillJellyFrameState {
    baseScaleX: SharedNumber;
    baseScaleXRate: SharedNumber;
    baseScaleY: SharedNumber;
    baseScaleYRate: SharedNumber;
    filteredVelocity: SharedNumber;
    filteredVelocityRate: SharedNumber;
    isDragging: SharedNumber;
    pressProgress: SharedNumber;
    pressProgressRate: SharedNumber;
    pressTarget: SharedNumber;
    rawPanelOffset: SharedNumber;
    rawPanelOffsetVelocity: SharedNumber;
    releasePending: SharedNumber;
    shapeTarget: SharedNumber;
    targetValue: SharedNumber;
    value: SharedNumber;
    valueVelocity: SharedNumber;
}

export interface PillJellyFrameConfig {
    releaseDistanceFraction: number;
    springs: {
        panel: SpringConfig;
        press: SpringConfig;
        scaleX: SpringConfig;
        scaleY: SpringConfig;
        value: SpringConfig;
        velocity: SpringConfig;
    };
}

const advancePosition = (
    state: PillJellyFrameState,
    config: PillJellyFrameConfig,
    deltaSeconds: number,
) => {
    "worklet";

    advanceSharedSpring(
        state.value,
        state.valueVelocity,
        state.targetValue.value,
        config.springs.value,
        deltaSeconds,
    );
};

const clampTargetValue = (
    state: PillJellyFrameState,
    tabCount: number,
) => {
    "worklet";

    state.targetValue.value = Math.min(
        Math.max(state.targetValue.value, 0),
        getMaxTabIndex(tabCount),
    );
};

const advanceFilteredVelocity = (
    state: PillJellyFrameState,
    config: PillJellyFrameConfig,
    tabCount: number,
    deltaSeconds: number,
) => {
    "worklet";

    const maxTabIndex = getMaxTabIndex(tabCount);
    const target =
        state.isDragging.value === 1 && maxTabIndex > 0
            ? state.valueVelocity.value / maxTabIndex
            : 0;

    advanceSharedSpring(
        state.filteredVelocity,
        state.filteredVelocityRate,
        target,
        config.springs.velocity,
        deltaSeconds,
    );
};

const advancePanelReturn = (
    state: PillJellyFrameState,
    config: PillJellyFrameConfig,
    deltaSeconds: number,
) => {
    "worklet";

    if (state.isDragging.value === 1) {
        return;
    }

    advanceSharedSpring(
        state.rawPanelOffset,
        state.rawPanelOffsetVelocity,
        0,
        config.springs.panel,
        deltaSeconds,
    );
};

const settleReleasedIndicator = (
    state: PillJellyFrameState,
    config: PillJellyFrameConfig,
    tabCount: number,
) => {
    "worklet";

    if (state.releasePending.value !== 1) {
        return;
    }

    const releaseDistance =
        Math.max(1, getMaxTabIndex(tabCount)) *
        config.releaseDistanceFraction;
    if (
        Math.abs(state.value.value - state.targetValue.value) >=
        releaseDistance
    ) {
        return;
    }

    state.releasePending.value = 0;
    state.pressTarget.value = 0;
    state.shapeTarget.value = 1;
};

const advancePress = (
    state: PillJellyFrameState,
    config: PillJellyFrameConfig,
    deltaSeconds: number,
) => {
    "worklet";

    advanceSharedSpring(
        state.pressProgress,
        state.pressProgressRate,
        state.pressTarget.value,
        config.springs.press,
        deltaSeconds,
    );
};

const advanceShape = (
    state: PillJellyFrameState,
    config: PillJellyFrameConfig,
    deltaSeconds: number,
) => {
    "worklet";

    const target = state.shapeTarget.value;
    advanceSharedSpring(
        state.baseScaleX,
        state.baseScaleXRate,
        target,
        config.springs.scaleX,
        deltaSeconds,
    );
    advanceSharedSpring(
        state.baseScaleY,
        state.baseScaleYRate,
        target,
        config.springs.scaleY,
        deltaSeconds,
    );
};

export const advancePillJellyFrame = (
    state: PillJellyFrameState,
    config: PillJellyFrameConfig,
    tabCount: number,
    timeSincePreviousFrame: number | null,
) => {
    "worklet";

    const deltaSeconds = getFrameDeltaSeconds(timeSincePreviousFrame);
    if (deltaSeconds === null) {
        return;
    }

    clampTargetValue(state, tabCount);
    advancePosition(state, config, deltaSeconds);
    advanceFilteredVelocity(state, config, tabCount, deltaSeconds);
    advancePanelReturn(state, config, deltaSeconds);
    settleReleasedIndicator(state, config, tabCount);
    advancePress(state, config, deltaSeconds);
    advanceShape(state, config, deltaSeconds);
};
